src/Native/System.Net.Security.Native/pal_ntlmapi.cpp

```diff
@@ -0,0 +1,267 @@
+    HMAC_Update(&ctx, static_cast<uint8_t*>(input), inputlen);
+    uint32_t hashLength;
+    HMAC_Final(&ctx, output, &hashLength);
+    HMAC_CTX_cleanup(&ctx);
+    return output;
+}
+
+static uint8_t* NetSecurity_EVPEncrypt(uint8_t* key, void* input, size_t inputlen)
+{
+    EVP_CIPHER_CTX ctx;
+    EVP_CIPHER_CTX_init(&ctx);
+    EVP_CipherInit_ex(&ctx, EVP_rc4(), nullptr, key, nullptr, 1);
+
+    uint8_t* output = new uint8_t[inputlen];
+    EVP_Cipher(&ctx, output, static_cast<uint8_t*>(input), static_cast<uint32_t>(inputlen));
+
+    EVP_CIPHER_CTX_cleanup(&ctx);
+    return output;
+}
+
+static int32_t
+NetSecurity_build_ntlm2_master(uint8_t* key, int32_t keylen, ntlm_buf* blob, ntlm_buf* sessionKey, ntlm_buf* masterKey)
+{
+    // reference: https://msdn.microsoft.com/en-us/library/cc236709.aspx
+    uint8_t* ntlmv2hash = NetSecurity_HMACDigest(key, keylen, blob->data, blob->length);
+    int32_t status = heim_ntlm_build_ntlm1_master(ntlmv2hash, UnsignedCast(keylen), sessionKey, masterKey);
+    if (status)
+    {
+        delete[] ntlmv2hash;
+        return status;
+    }
+
+    uint8_t* exportKey = NetSecurity_EVPEncrypt(ntlmv2hash, sessionKey->data, sessionKey->length);
+    delete[] ntlmv2hash;
+    masterKey->length = sessionKey->length;
+    masterKey->data = exportKey;
+    return status;
+}
+
+extern "C" int32_t NetSecurity_CreateType3Message(ntlm_buf* key,
+                                                  ntlm_type2* type2,
+                                                  const char* username,
+                                                  const char* domain,
+                                                  uint32_t flags,
+                                                  ntlm_buf* lmResponse,
+                                                  ntlm_buf* ntlmResponse,
+                                                  uint8_t* baseSessionKey,
+                                                  int32_t baseSessionKeyLen,
+                                                  ntlm_buf** outSessionHandle,
+                                                  int* outSessionKeyLen,
+                                                  ntlm_buf** outBufferHandle,
+                                                  int32_t* outLength)
+{
+    assert(key != nullptr);
+    assert(type2 != nullptr);
+    assert(lmResponse != nullptr);
+    assert(ntlmResponse != nullptr);
+    static char* workstation = static_cast<char*>(calloc(1, sizeof(char))); // empty string
+    ntlm_type3 type3;
+    memset(&type3, 0, sizeof(ntlm_type3));
+    type3.username = const_cast<char*>(username);
+    type3.targetname = const_cast<char*>(domain);
+    type3.lm = *lmResponse;
+    type3.ntlm = *ntlmResponse;
+    type3.ws = workstation;
+    type3.flags = flags;
+
+    int32_t status = 0;
+    ntlm_buf masterKey = {.length = 0, .data = nullptr};
+    *outSessionHandle = new ntlm_buf();
+
+    if (type2->targetinfo.length == 0)
```

Comment:
Rather than repeating this check in this function save the value to a local.  That way if it ends up changing during the function execution we don't free in the wrong way.